PowerEditor/src/DarkMode/DarkMode.h

```diff
@@ -0,0 +1,271 @@
+{
+	HIGHCONTRASTW highContrast = { sizeof(highContrast) };
+	if (SystemParametersInfoW(SPI_GETHIGHCONTRAST, sizeof(highContrast), &highContrast, FALSE))
+		return highContrast.dwFlags & HCF_HIGHCONTRASTON;
+	return false;
+}
+
+void RefreshTitleBarThemeColor(HWND hWnd)
+{
+	BOOL dark = FALSE;
+	if (_IsDarkModeAllowedForWindow(hWnd) &&
+		_ShouldAppsUseDarkMode() &&
+		!IsHighContrast())
+	{
+		dark = TRUE;
+	}
+	if (g_buildNumber < 18362)
+		SetPropW(hWnd, L"UseImmersiveDarkModeColors", reinterpret_cast<HANDLE>(static_cast<INT_PTR>(dark)));
+	else if (_SetWindowCompositionAttribute)
+	{
+		WINDOWCOMPOSITIONATTRIBDATA data = { WCA_USEDARKMODECOLORS, &dark, sizeof(dark) };
+		_SetWindowCompositionAttribute(hWnd, &data);
+	}
+}
+
+bool IsColorSchemeChangeMessage(LPARAM lParam)
+{
+	bool is = false;
+	if (lParam && (0 == lstrcmpi(reinterpret_cast<LPCWCH>(lParam), L"ImmersiveColorSet")))
+	{
+		_RefreshImmersiveColorPolicyState();
+		is = true;
+	}
+	_GetIsImmersiveColorUsingHighContrast(IHCM_REFRESH);
+	return is;
+}
+
+bool IsColorSchemeChangeMessage(UINT message, LPARAM lParam)
+{
+	if (message == WM_SETTINGCHANGE)
+		return IsColorSchemeChangeMessage(lParam);
+	return false;
+}
+
+void AllowDarkModeForApp(bool allow)
+{
+	if (_AllowDarkModeForApp)
+		_AllowDarkModeForApp(allow);
+	else if (_SetPreferredAppMode)
+		_SetPreferredAppMode(allow ? AllowDark : Default);
+}
+
+// limit dark scroll bar to specific windows and their children
+
+std::unordered_set<HWND> g_darkScrollBarWindows;
+std::mutex g_darkScrollBarMutex;
+
+void EnableDarkScrollBarForWindowAndChildren(HWND hwnd)
+{
+	std::lock_guard<std::mutex> lock(g_darkScrollBarMutex);
+	g_darkScrollBarWindows.insert(hwnd);
+}
+
+bool IsWindowOrParentUsingDarkScrollBar(HWND hwnd)
+{
+	HWND hwndRoot = GetAncestor(hwnd, GA_ROOT);
+
+	std::lock_guard<std::mutex> lock(g_darkScrollBarMutex);
+	if (g_darkScrollBarWindows.count(hwnd)) {
+		return true;
+	}
+	if (hwnd != hwndRoot && g_darkScrollBarWindows.count(hwndRoot)) {
+		return true;
+	}
+
+	return false;
+}
+
+void FixDarkScrollBar()
+{
+	HMODULE hComctl = LoadLibraryExW(L"comctl32.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
+	if (hComctl)
+	{
+		auto addr = FindDelayLoadThunkInModule(hComctl, "uxtheme.dll", 49); // OpenNcThemeData
+		if (addr)
+		{
+			DWORD oldProtect;
+			if (VirtualProtect(addr, sizeof(IMAGE_THUNK_DATA), PAGE_READWRITE, &oldProtect))
+			{
+				auto MyOpenThemeData = [](HWND hWnd, LPCWSTR classList) -> HTHEME {
+					if (wcscmp(classList, L"ScrollBar") == 0)
+					{
+						if (IsWindowOrParentUsingDarkScrollBar(hWnd)) {
+							hWnd = nullptr;
+							classList = L"Explorer::ScrollBar";
+						}
+					}
+					return _OpenNcThemeData(hWnd, classList);
+				};
+
+				addr->u1.Function = reinterpret_cast<ULONG_PTR>(static_cast<fnOpenNcThemeData>(MyOpenThemeData));
+				VirtualProtect(addr, sizeof(IMAGE_THUNK_DATA), oldProtect, &oldProtect);
+			}
+		}
+	}
+}
+
+constexpr bool CheckBuildNumber(DWORD buildNumber)
```

Comment:
A hardcoded build list? What shall we do for the new builds? Just add another hardcoded value?